File: c++/1127.cpp
//
//  main.cpp
//  test
//
//  Created by 李寻欢 on 2019/10/15.
//  Copyright © 2019 李寻欢. All rights reserved.
//
#include <stdio.h>
#include <iostream>
#include <string>
#include <map>
#include <cmath>
#include <algorithm>
#include <stack>
#include <vector>
#include <queue>
#include <unordered_map>
#include <unordered_set>
#include <string.h>
#include <set>
#include <deque>
using namespace std;
const int MAXN = 40;
int n;
int Inorder[MAXN];
int Postorder[MAXN];

typedef struct node{
    int val;
    node* left;
    node* right;
    node(int x){
        val = x;
        left = right = NULL;
    }
}Node,*TreeNode;

TreeNode construct(int inL,int inR,int postL,int postR){
    if(inL > inR) return NULL;
    int k = -1;
    for(int i=inL;i<=inR;++i){
        if(Inorder[i] == Postorder[postR]){
            k = i;
            break;
        }
    }
    TreeNode root = new node(Postorder[postR]);
    int numleft = k-inL;
    root->left = construct(inL, inL+numleft-1, postL, postL+numleft-1);
    root->right = construct(k+1, inR, postL+numleft, postR-1);
    return root;
}

void level(TreeNode root){
    queue<TreeNode> q;
    q.push(root);
    int flag = -1;
    int space = 0;
    while (q.size()) {
        int size = (int)q.size();
        flag *= -1;
        vector<int> v;
        while (size--) {
            TreeNode cur = q.front();
            q.pop();
            v.emplace_back(cur->val);
            if(cur->left) q.push(cur->left);
            if(cur->right) q.push(cur->right);
        }
        if(flag == -1){
            for(int i=0;i<v.size();++i){
                printf("%d",v[i]);
                space++;
                if(space != n) printf(" ");
            }
        }else{
            for(int i=(int)v.size()-1;i>=0;--i){
                printf("%d",v[i]);
                space++;
                if(space != n) printf(" ");
            }
        }
    }
}

int main(){
    scanf("%d",&n);
    for(int i=0;i<n;++i) scanf("%d",&Inorder[i]);
    for(int i=0;i<n;++i) scanf("%d",&Postorder[i]);
    TreeNode root = construct(0, n-1, 0, n-1);
    level(root);
}

File: Python3/1127.py
from queue import Queue
n=int(input())
Inorder = list(map(int,input().split()))
Postorder = list(map(int,input().split()))
class node:
    def __init__(self,val):
        self.val = val
        self.left = None
        self.right = None
def construct(inl,inr,postl,postr):
    if inl > inr:
        return None
    k = Inorder.index(Postorder[postr])
    root = node(Postorder[postr])
    numleft = k-inl
    root.left = construct(inl,inl+numleft-1,postl,postl+numleft-1)
    root.right = construct(k+1,inr,postl+numleft,postr-1)
    return root
root = construct(0,n-1,0,n-1)
q = Queue()
q.put(root)
flag = 1
space = 0
while q.qsize():
    size = q.qsize()
    now = []
    while size:
        size -= 1
        cur = q.get()
        now.append(cur.val)
        if cur.left:
            q.put(cur.left)
        if cur.right:
            q.put(cur.right)
    if flag == -1:
        for each in now:
            space += 1
            print(each,end='')
            if space != n:  print(' ',end='')
    else:
        for each in now[::-1]:
            space += 1
            print(each,end='')
            if space != n:  print(' ',end='')
    flag *= -1
